### ISSUE TITLE
Borrow checker incorrectly accepts code when calling function with complex lifetime bounds through a fn pointer

### ISSUE THREAD

**Issue by ennis** [rust-lang/rust#57170]
The following code compiles and prints a bogus value: 
```rust
fn shorten_lifetime<'a, 'b, 'min>(a: &'a i32, b: &'b i32) -> &'min i32
where 'a: 'min, 'b: 'min
{
    if *a < *b {
        &a
    } else {
        &b
    }
}


fn main()
{
    let fnptr = &shorten_lifetime;
    
    let a = &5;
    let ptr = {
        let b = 3;
        let b = &b;
        //let c = shorten_lifetime(a, b);
        let c = fnptr(a, b);
        c
    };
    
    println!("ptr = {:?}", ptr);
    
}
```

Output (release mode):
```
ptr = -1876488848
```
 
(playground: https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=048d856b4689599b1e35f0a44ebe629f)

Expected result: a borrow checker error, the same as the one obtained when calling `shorten_lifetime` directly.

I accidentally ran into this issue when trying to spell the fn pointer type of `fnptr` (I'm not sure how to add constraints to the lifetimes of a higher-ranked fn type like `for<'a, 'b> fn(...) -> T`)

**Comment by jonas-schievink** [comment]
The 2015 edition correctly rejects this (which doesn't yet have NLL)

**Comment by matthewjasper** [comment]
It appears that we lose the function bounds when we promote it. Relavant FIXME: https://github.com/rust-lang/rust/blob/48742c68d9e53fba6b48c211140db595fbaa491e/src/librustc_mir/borrow_check/nll/type_check/mod.rs#L433-L443

**Comment by pnkfelix** [comment]
triage: P-high. Hopefully fix in PR #57202 lands very soon.
